### PR TITLE
Add --include-dotfiles option to babel-cli

### LINKED ISSUE #5052: --copy-files doesn't copy dot-files
### Input Code

```js
// A folder with dot-files (from dotenv NPM)
config/.env
config/.env.required
```

### Babel Configuration (.babelrc, package.json, cli command)

```js
{
  "presets": [
    "es2015"
  ]
}

./node_modules/babel-cli/bin/babel.js functions -d build/ --copy-files --source-maps
```

### Expected Behavior

When using --copy-files dot-files and folders that contain them should be copied as well

### Current Behavior

Folders and dot-files are not copied

### Possible Solution

Files should be copied by default or provide an additional flag or config to include or exclude copying of dot-files

### Context

Wasn't able to run transpiled code without adding a post step to copy the dot-files.

### Your Environment

| software         | version
| ---------------- | -------
| Babel            |  6.18.0
| node             | 4.5.0
| npm              | 3.10.8
| Operating System | macOS 10.12.2

### PR DESCRIPTION
**Pull request by existentialism**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues             | Fixes #5052, Closes #5083
| Patch: Bug Fix?          | n
| Major: Breaking Change?  | n
| Minor: New Feature?      | y
| Tests Added/Pass?        | y/y
| Spec Compliancy?         | n
| License                  | MIT
| Doc PR                   | 
| Any Dependency Changes?  | 

Picking https://github.com/babel/babel/pull/5083 back up.

I'll also update the docs after this lands.

### CODE REVIEW

**Comment by Swivelgames** [comment]
👍

File: packages/babel-cli/src/babel/dir.js [hunk 1]
```diff
@@ -71,7 +71,14 @@ export default function(commander, filenames, opts) {
       if (commander.deleteDirOnStart) {
         util.deleteDir(commander.outDir);
       }
-      util.readdir(dirname).forEach(function(filename) {
+
+      let readDirFilter;
+
+      if (commander.copyDotFiles) {
+        readDirFilter = () => true;
```
Comment (Swivelgames) [inline packages/babel-cli/src/babel/dir.js]: @existentialism Does this also ignore `.` and `..`? Would `(x) => x !== '.' && x !== '..'` work better? I'm sure you've done the tests and all that, so I'm just curious. 👍

File: packages/babel-cli/src/babel/index.js [hunk 2]
```diff
@@ -154,10 +154,14 @@ commander.option(
   "-D, --copy-files",
   "When compiling a directory copy over non-compilable files",
 );
+commander.option(
+  "--copy-dot-files",
+  "When --copy-files is set, dot files are copied as well",
```
Comment (mathiasbynens) [inline packages/babel-cli/src/babel/index.js]: Super minor nit, but “dotfiles” is generally written as a single word. Can we change the option to `--copy-dotfiles` and `s/dot files/dotfiles/g`?

File: packages/babel-cli/src/babel/index.js [hunk 3]
```diff
@@ -233,6 +241,7 @@ delete opts.skipInitialBuild;
 delete opts.outFile;
 delete opts.outDir;
 delete opts.copyFiles;
+delete opts.copyDotFiles;
```
Comment (mathiasbynens) [inline packages/babel-cli/src/babel/index.js]: Then this would become `copyDotfiles`.

File: packages/babel-cli/src/babel/index.js [hunk 4]
```diff
@@ -212,6 +216,10 @@ if (commander.deleteDirOnStart && !commander.outDir) {
   errors.push("--delete-dir-on-start requires --out-dir");
 }
 
+if (commander.copyDotfiles && !commander.copyFiles) {
+  errors.push("--copy-dotfiles requires --copy-files");
```
Comment (loganfsmyth) [inline packages/babel-cli/src/babel/index.js]: Is this strictly true? I guess technically you could have `.foo.js` files that you still expect to be compiled and thus don't need dotFiles for.

Maybe what we should do is call it `--include-dotfiles` so if there are dotfiles they get included in whatever operation?

**Comment by existentialism** [comment]
Updated as `--include-dotfiles` which includes them when compiling _and_ copying files.

**Pull request merged**